ethcore/src/chainfilter/chainfilter.rs

```diff
@@ -0,0 +1,199 @@
+			// map them to offsets
+			.map(|li| li.index * level_size)
+			// get all blocks that may contain our bloom
+			.map(|off| self.blocks(bloom, from_block, to_block, level - 1, off))
+			// filter existing ones
+			.filter_map(|x| x)
+			// flatten nested structures
+			.flat_map(|v| v)
+			.collect();
+		Some(res)
+	}
+
+	/// Adds new bloom to all filter levels
+	pub fn add_bloom(&self, bloom: &H2048, block_number: usize) -> HashMap<BloomIndex, H2048> {
+		let mut result: HashMap<BloomIndex, H2048> = HashMap::new();
+
+		for level in 0..self.indexer.levels() {
+			let bloom_index = self.indexer.bloom_index(block_number, level);
+			let new_bloom = match self.data_source.bloom_at_index(&bloom_index) {
+				Some(old_bloom) => old_bloom | bloom.clone(),
+				None => bloom.clone(),
+			};
+
+			result.insert(bloom_index, new_bloom);
+		}
+
+		result
+	}
+
+	/// Resets blooms at level 0 and forces rebuild on higher levels.
+	pub fn reset_chain_head(&self, blooms: &[H2048], block_number: usize, old_highest_block: usize) -> HashMap<BloomIndex, H2048> {
+		let mut result: HashMap<BloomIndex, H2048> = HashMap::new();
+
+		// insert all new blooms at level 0
+		for (i, bloom) in blooms.iter().enumerate() {
+			result.insert(self.indexer.bloom_index(block_number + i, 0), bloom.clone());
+		}
+
+		// reset the rest of blooms
+		for reset_number in block_number + blooms.len()..old_highest_block {
+			result.insert(self.indexer.bloom_index(reset_number, 0), H2048::new());
+		}
+
+		for level in 1..self.indexer.levels() {
+			for i in 0..blooms.len() {
+
+				let index = self.indexer.bloom_index(block_number + i, level);
+				let new_bloom = {	
+					// use new blooms before db blooms where necessary
+					let bloom_at = | index | { result.get(&index).cloned().or_else(|| self.data_source.bloom_at_index(&index)) };
+
+					self.indexer.lower_level_bloom_indexes(&index)
+						.into_iter()
+						// get blooms
+						.map(bloom_at)
+						// filter existing ones
+						.filter_map(|b| b)
+						// BitOr all of them
+						.fold(H2048::new(), |acc, bloom| acc | bloom)
+				};
+
+				result.insert(index, new_bloom);
+			}
+		}
+
+		result
+	}
+
+	/// Returns numbers of blocks that may log bloom.
+	pub fn blocks_with_bloom(&self, bloom: &H2048, from_block: usize, to_block: usize) -> Vec<usize> {
+		let mut result = vec![];
+		// lets start from highest level
+		let max_level = self.indexer.max_level();
+		let level_size = self.indexer.level_size(max_level);
+		let from_index = self.indexer.bloom_index(from_block, max_level);
+		let to_index = self.indexer.bloom_index(to_block, max_level);
+
+		for index in from_index.index..to_index.index + 1 {
+			// offset will be used to calculate where we are right now
+			let offset = level_size * index;
+
+			// go doooown!
+			if let Some(blocks) = self.blocks(bloom, from_block, to_block, max_level, offset) {
+				result.extend(blocks);
+			}
+		}
+
+		result
+	}
+}
```

Comment:
tests are in a separate file: https://github.com/ethcore/parity/pull/418/files#diff-adb08cad471cefdc2a84e4901a06a48aR1